storage: fix tombstone handling in applySnapshot

tbg opened the pull request:
`applySnapshot` had no special-casing of the unintentionally replicated
legacy Raft tombstone key. As a consequence of this, applying a snapshot
would wipe any existing tombstone. This must have been a bug for a long
time, but it became painfully obvious after #21120, which migrates the
Raft tombstone to the unreplicated keyspace, and in the process added
an explicit removal of the legacy tombstone during snapshot application
plus corresponding testing (which caught this bug).

This commit makes sure that when a snapshot is applied, any legacy
tombstone is propagated (i.e. preserved), and removes some unrelated
flakes in the failing test: the test was laying down tombstones
without synchronizing with the running nodes, prompting Raft panics
(due to rewinding the tombstone's NextReplicaID); it was also not
taking into account a situation in which a legacy tombstone could
actually be rewritten into a new-style tombstone (after replicaGC).

Fixes #21367.

Release note(bug fix): fixed a bug that could lead to a crash (which in
turn would require that the node be reinitialized).

cockroach-teamcity commented:
<!-- Reviewable:start -->
This change is [<img src="https://reviewable.io/review_button.png" height="34" align="absmiddle" alt="Reviewable"/>](https://reviewable.io/reviews/cockroachdb/cockroach/21771)
<!-- Reviewable:end -->

tbg committed:
storage: fix tombstone handling in applySnapshot

`applySnapshot` had no special-casing of the unintentionally replicated
legacy Raft tombstone key. As a consequence of this, applying a snapshot
would wipe any existing tombstone. This must have been a bug for a long
time, but it became painfully obvious after #21120, which migrates the
Raft tombstone to the unreplicated keyspace, and in the process added
an explicit removal of the legacy tombstone during snapshot application
plus corresponding testing (which caught this bug).

This commit makes sure that when a snapshot is applied, any legacy
tombstone is propagated (i.e. preserved), and removes some unrelated
flakes in the failing test: the test was laying down tombstones
without synchronizing with the running nodes, prompting Raft panics
(due to rewinding the tombstone's NextReplicaID); it was also not
taking into account a situation in which a legacy tombstone could
actually be rewritten into a new-style tombstone (after replicaGC).

Fixes #21367.

Release note(bug fix): fixed a bug that could lead to a crash (which in
turn would require that the node be reinitialized).

bdarnell commented:
<img class="emoji" title=":lgtm:" alt=":lgtm:" align="absmiddle" src="https://reviewable.io/lgtm.png" height="20" width="61"/>

---

Reviewed 4 of 4 files at r1.
Review status: all files reviewed at latest revision, all discussions resolved, some commit checks pending.

---



*Comments from [Reviewable](https://reviewable.io:443/reviews/cockroachdb/cockroach/21771#-:-L3eSxDrniuJKXf6WvTn:bnfp4nl)*
<!-- Sent from Reviewable.io -->

tschottdorf commented:
TFTR!
